pkg/userdata/centos/provider.go

```diff
@@ -334,6 +337,42 @@ write_files:
   append: true
 {{- end }}
 
+{{- if eq .CloudProviderName "kubevirt" }}
+- path: "/opt/bin/restart-kubelet.sh"
+  permissions: "0744"
+  content: |
+    #!/bin/bash
+    # Needed for Kubevirt provider because if the virt-launcher pod is deleted,
+    # the VM and DataVolume states are kept and VM is rebooted. We need to restart the kubelet
+    # with the new config (new IP) and run this at every boot.
+    set -xeuo pipefail
+
+    # This helps us avoid an unnecessary restart for kubelet on the first boot
+    if [ -f /etc/kubelet_needs_restart ]; then
+      # restart kubelet since it's not the first boot
+      systemctl daemon-reload
+      systemctl restart kubelet.service
+    else
+      touch /etc/kubelet_needs_restart
+    fi
```

Comment:
I'm just randomly stumbling into this PR but I'm curious, how is this script different from the systemd service `kubelet.service` being enabled in line 247 (https://github.com/kubermatic/machine-controller/pull/1246/files#diff-ad511f10cb3024c5aa13e897590c6bc85d1ba948341c05da6d3c6ac9dde196e0R247)? Shouldn't systemd start the service?

Reply:
My point is that the `kubelet` service isn't started, it's set up via `systemctl enable --now`:

https://github.com/kubermatic/machine-controller/blob/b6567d9a57c5d88984401367c2e952db828ac9e3/pkg/userdata/centos/provider.go#L247

That both starts and _enables_ the service. In systemd, "enabled" means it will start it at system startup/boot.

Reply:
Is it because the service fails after a reboot? That might mean we have a dependency issue in the service definition.

Reply:
@hdurand0710 ^^

Reply:
it did not fail at reboot but was started with the wrong configuration. We needed to reload the configuration and restart.

Reply:
Sorry for being such a PITA, I just don't think I understand it yet. How can it run with the wrong configuration after the system just freshly booted up? Is there a systemd service that refreshes a config file that kubelet relies on? If yes, couldn't we just add a dependency for the kubelet service on that service when on kubevirt?

Reply:
The script reload systemd which makes me think that the kubelet.service systemd unit itself changed, but what is doing that?

Reply:
The configuration is updated in /opt/bin/setup_net_env.sh.
After reboot, this file was correct : /etc/systemd/system/kubelet.service.d/nodeip.conf
Nut the kubelet was running with the previous configuration.
So the idea was to reload the configuration and restart the kubelet.
If you think there a better way, let's go for it.

Reply:
Yeah okay, dealing with updated systemd units is painful. Thanks for explaining it to me, I really didn't get it. Sorry for wasting your time :-)

Reply:
Any feedback is never a waste of time. Always good! Thanks for taking the time to read this `systems` thingy. :-)